### PR TITLE
Add support for Hetzner Cloud Floating IPs 

### PR DESCRIPTION
**Pull request by christoph-heiss**

### CODE REVIEW

File: ipmanager/hetznerFloatingIpConfigurer.go
Comment (markwort) [inline ipmanager/hetznerFloatingIpConfigurer.go:49]: Checking once a minute would be a more sensible starting point. Otherwise, we need to make it configurable.
Looking at the old (hetzner robot api) code, we only bothered to recheck that once every hour.
This is really only a courtesy check, following the (successful) registration of the VIP we can expect it to continue pointing our way, unless somebody tampers with it from the outside, in which case it is mainly their responsibility to fix, not ours.

File: ipmanager/hetznerFloatingIpConfigurer.go
Comment (markwort) [inline ipmanager/hetznerFloatingIpConfigurer.go:53]: This should only be printed when verbose logs are required as well...

File: ipmanager/hetznerFloatingIpConfigurer.go
Comment (markwort) [inline ipmanager/hetznerFloatingIpConfigurer.go:33]: Please put this into its own method so that we can recreate this whenever access to the API fails.

File: ipmanager/hetznerFloatingIpConfigurer.go
Comment (markwort) [inline ipmanager/hetznerFloatingIpConfigurer.go:102]: These requests also count against the rate limit.
Can we cache the `floatingIp` in the `HetznerFloatingIpConfigurer` and update it only whenever the cache is too old (i.e. in the query function) ?

File: ipmanager/hetznerFloatingIpConfigurer.go
Comment (markwort) [inline ipmanager/hetznerFloatingIpConfigurer.go:117]: Can we add a timeout to this select as well? We don't want to wait indefinitely here, or does the hetzner library already catch timeouts and gives us an error after a certain time?

File: ipmanager/hetznerFloatingIpConfigurer.go
Comment (markwort) [inline ipmanager/hetznerFloatingIpConfigurer.go:149]: (same as in the "configure" function):


>These requests also count against the rate limit.
>Can we cache the floatingIp in the HetznerFloatingIpConfigurer and update it only whenever the cache is too old (i.e. in the query function) ?

File: ipmanager/hetznerFloatingIpConfigurer.go
Comment (markwort) [inline ipmanager/hetznerFloatingIpConfigurer.go:164]: (same as in the "configure" function):
> Can we add a timeout to this select as well? We don't want to wait indefinitely here, or does the hetzner library already catch timeouts and gives us an error after a certain time?

File: ipmanager/ip_manager.go
Comment (markwort) [inline ipmanager/ip_manager.go:31]: If it is not too complicated, we can get rid of the `IPConfiguration` struct altogether, I don't think there is much that the `vipconfig.Config` doesn't already contain or could be easily made to contain.

File: test/hetzner_floating_ip_test.sh
Comment (markwort) [inline test/hetzner_floating_ip_test.sh:10]: don't think we need that here :)

File: vipconfig/vip-manager.yml
Comment (markwort) [inline vipconfig/vip-manager.yml:16]: Please add an oxford comma:
 `basic`, `hetzner`, or `hetzner_floating_ip`

**Review by markwort — CHANGES_REQUESTED** [review]
Dear @christoph-heiss , please make some small tweaks. I think then this is ready to ship.
Good job!

File: README.md
Comment (christoph-heiss) [reply README.md]: I also think `hetzner-cloud` is better and renaming everything from Floating IP to Hetzner Cloud is very senseful, so I've implemented that variant.

File: ipmanager/ip_manager.go
Comment (christoph-heiss) [reply ipmanager/ip_manager.go:31]: So I've tried to remove `IPConfiguration` completely, but as it turns out this is way more work than really justifiable for this pull request. 
If we really want to clean that up, it could be done as a separate pull request.

File: test/hetzner_floating_ip_test.sh
Comment (christoph-heiss) [reply test/hetzner_floating_ip_test.sh]: Originally I've put this in place since `ip`, `netmask` and `interface` must be configured/passed correctly for vip-manager to work, since it will query the network interface on startup.
I worked around this in commit 629c1bd4c235d4ca7c201cb1440d11b481cbc865, hope this solution works for you too.

File: ipmanager/hetznerFloatingIpConfigurer.go
Comment (christoph-heiss) [reply ipmanager/hetznerFloatingIpConfigurer.go]: Since the `hcloud` library performs HTTP request against their publicly documented API, the request will eventually time out if something is very wrong and the action will send an error on the `errCh`.

**Comment by pashagolub** [comment]
Hey @tpo 

We don't have a dedicated maintainer for the project, and that's probably why all this new functionality is still unmerged.

That said, I'm always trying to support one's work. So yeah, I would be more than happy to work with perfect pull requests. :-)

I'm not sure we have any projects involving hetzner, meaning we cannot test that in the field. But I'm OK with having one more method. @markwort what do you think?

**Comment by christoph-heiss** [comment]
Chiming in:
We do have a test setup (read: one Hetzner VPS + floating IP) - so yes, we can test this functionality, at least for development.

I can clean up this PR, rebasing and resolving the remaining comments/CI warnings as I got time. Shouldn't be much of a hassle, since (IIRC) everything was already working as intended.
Other than that, I think it as mergable as is - @markwort?

**Comment by pashagolub** [comment]
Good to know @christoph-heiss. Feel free to reopen the PR then. 👍

**Comment by tpo** [comment]
> We don't have a dedicated maintainer for the project, and that's probably why all this new functionality is still unmerged.
> 
> That said, I'm always trying to support one's work. So yeah, I would be more than happy to work with perfect pull requests. :-)
> 
> I'm not sure we have any projects involving hetzner, meaning we cannot test that in the field. But I'm OK with having one more method. @markwort what do you think?

All right, very nice to hear and thanks a lot @pashagolub ! So in case any one of you would **very much still want** a code review from me, then please let me know and I'll try to do it earlier that one year from now ;-D. And I'll also wait for a "yes I want this patch merged" from @markwort .

Thanks a lot to you guys for caring about this!

**Pull request closed without merging**
